widgetssdk/src/main/java/com/glia/widgets/core/dialog/DialogController.kt

```diff
@@ -0,0 +1,177 @@
+    private val viewCallbacks: MutableSet<DialogController.Callback> = HashSet()
+    private val dialogManager: DialogManager by lazy { DialogManager(::emitDialogState) }
+    override val isShowingUnexpectedErrorDialog: Boolean
+        get() = dialogManager.currentDialogState is DialogState.UnexpectedError
+
+    override val isEnableScreenSharingNotificationsAndStartSharingDialogShown: Boolean
+        get() = DialogState.EnableScreenSharingNotificationsAndStartSharing == dialogManager.currentDialogState
+    private val isOverlayDialogShown: Boolean
+        get() = DialogState.OverlayPermission == dialogManager.currentDialogState
+
+    override fun dismissCurrentDialog() {
+        Logger.d(TAG, "Dismiss current dialog")
+        dialogManager.dismissCurrent()
+    }
+
+    override fun dismissDialogs() {
+        Logger.d(TAG, "Dismiss dialogs")
+        dialogManager.dismissAll()
+    }
+
+    private fun emitDialogState(dialogState: DialogState) {
+        Logger.d(TAG, "Emit dialog state:\n$dialogState")
+        viewCallbacks.forEach { it.emitDialogState(dialogState) }
+    }
+
+    override fun showExitQueueDialog() {
+        Logger.i(TAG, "Show Exit Queue Dialog")
+        dialogManager.addAndEmit(DialogState.ExitQueue)
+    }
+
+    override fun showExitChatDialog() {
+        Logger.i(TAG, "Show End Engagement Dialog")
+        dialogManager.addAndEmit(DialogState.EndEngagement)
+    }
+
+    override fun showUpgradeAudioDialog(mediaUpgradeOffer: MediaUpgradeOffer, operatorName: String?) {
+        Logger.i(TAG, "Show Upgrade Audio Dialog")
+        dialogManager.addAndEmit(DialogState.MediaUpgrade(mediaUpgradeOffer, operatorName, MediaUpgradeMode.AUDIO))
+    }
+
+    override fun showUpgradeVideoDialog2Way(mediaUpgradeOffer: MediaUpgradeOffer, operatorName: String?) {
+        Logger.i(TAG, "Show Upgrade 2WayVideo Dialog")
+        dialogManager.addAndEmit(DialogState.MediaUpgrade(mediaUpgradeOffer, operatorName, MediaUpgradeMode.VIDEO_TWO_WAY))
+    }
+
+    override fun showUpgradeVideoDialog1Way(mediaUpgradeOffer: MediaUpgradeOffer, operatorName: String?) {
+        Logger.i(TAG, "Show Upgrade 1WayVideo Dialog")
+        dialogManager.addAndEmit(DialogState.MediaUpgrade(mediaUpgradeOffer, operatorName, MediaUpgradeMode.VIDEO_ONE_WAY))
+    }
+
+    override fun showVisitorCodeDialog() {
+        Logger.i(TAG, "Show Visitor Code Dialog")
+        if (isOverlayDialogShown) {
+            dialogManager.add(DialogState.VisitorCode)
+        } else {
+            dialogManager.addAndEmit(DialogState.VisitorCode)
+        }
+    }
+
+    override fun dismissVisitorCodeDialog() {
+        Logger.i(TAG, "Dismiss Visitor Code Dialog")
+        dialogManager.remove(DialogState.VisitorCode)
+    }
+
+    override fun showUnexpectedErrorDialog() {
+        // PRIORITISE THIS ERROR AS IT IS ENGAGEMENT FATAL ERROR INDICATOR (eg. GliaException:{"details":"Queue is closed","error":"Unprocessable entity"}) for example
```

Comment:
What does this comment mean? Is it a TODO? Does it mean that some additional changes needed?

Reply:
This is just information about why we're doing so(prioritizing this dialog).

Reply:
Does it make sense to make it lowercase instead of uppercase?

Reply:
I don't know who wrote this comment, but I think it makes sense, and since I already merged this branch will do it on my next PR.

Reply:
Ah, no problem, we can leave it uppercase, it's not critical.